### PR TITLE
Add USB support for RT600 & LPC55S69

### PR DESCRIPTION
**Pull request by mmahadevan108**
Tested using `samples/subsys/usb/hid`,  `samples/subsys/usb/hid-mouse`, `samples/subsys/usb/mass`

### CODE REVIEW

**Comment by klaasdannen** [comment]
Why limited to those two MCUs? What about the LPC54S16 which is also part of Zephyr and that supports USB?

File: drivers/usb/device/usb_dc_mcux_lpcip3511.c [hunk 1]
```diff
@@ -0,0 +1,775 @@
+/*
+ * Copyright (c) 2018-2019, NXP
```
Comment (MaureenHelm) [inline drivers/usb/device/usb_dc_mcux_lpcip3511.c:2]: Update to include 2021

File: drivers/usb/device/usb_dc_mcux_lpcip3511.c [hunk 2]
```diff
@@ -0,0 +1,775 @@
...
+
+/* the setup transfer state */
+#define SETUP_DATA_STAGE_DONE	(0)
+#define SETUP_DATA_STAGE_IN	(1)
+#define SETUP_DATA_STAGE_OUT	(2)
+
+/*
+ * Endpoint absolute index calculation:
+ *
+ * MCUX EHCI USB device controller supports a specific
```
Comment (MaureenHelm) [inline drivers/usb/device/usb_dc_mcux_lpcip3511.c:37]: Update to LPCIP3511

File: drivers/usb/device/usb_dc_mcux_lpcip3511.c [hunk 3]
```diff
@@ -0,0 +1,775 @@
...
+K_MEM_SLAB_DEFINE(ep_buf_pool, 1024, EP_BUF_NUMOF_BLOCKS, 64);
+
+static usb_ep_ctrl_data_t s_ep_ctrl[NUM_OF_EP_MAX];
+static usb_device_struct_t dev_data;
+
+/* LPCIP3511 device driver interface */
+static const usb_device_controller_interface_struct_t ip3511 = {
+		USB_DeviceLpc3511IpInit, USB_DeviceLpc3511IpDeinit, USB_DeviceLpc3511IpSend,
+		USB_DeviceLpc3511IpRecv, USB_DeviceLpc3511IpCancel, USB_DeviceLpc3511IpControl
+};
```
Comment (MaureenHelm) [inline drivers/usb/device/usb_dc_mcux_lpcip3511.c:76]: Would this interface abstraction make it possible to merge this driver with the existing MCUX EHCI driver?

Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux_lpcip3511.c]: I have merged the two drivers. I have also renamed `usb_dc_mcux_ehci.c`  to `usb_dc_mcux.c` to reflect this change.

**Comment by mmahadevan108** [comment]
Any guidance on this build failure
```
HEAD is now at c74be389 trusted-firmware-m: platform: lpcxpresso55s69: Update SDK
ERROR: update failed for project lvgl

```

File: modules/hal_nxp/usb/usb_dc_mcux.h [hunk 4]
```diff
@@ -25,6 +25,43 @@
...
+#endif
+
+/* big/little endian */
+#define SWAP2BYTE_CONST(n) ((((n)&0x00FFU) << 8U) | (((n)&0xFF00U) >> 8U))
+
+#if defined(CONFIG_BIG_ENDIAN)
+#define USB_SHORT_FROM_LITTLE_ENDIAN(n) SWAP2BYTE_CONST(n)
+#else
+#define USB_SHORT_FROM_LITTLE_ENDIAN(n) (n)
+#endif
```
Comment (MaureenHelm) [inline modules/hal_nxp/usb/usb_dc_mcux.h:54]: Use helper functions in `include/sys/byteorder.h`

**Review by MaureenHelm — COMMENTED** [review]
This looks much better after merging with the EHCI driver and will be easier to maintain, thanks. I tested `samples/subsys/usb/hid-mouse` on `mimxrt1064_evk` and `mimxrt685_evk_cm33`.

> Any guidance on this build failure
>
> ```HEAD is now at c74be389 trusted-firmware-m: platform: lpcxpresso55s69: Update SDK ERROR: update failed for project lvgl```

No, but it looks like there are merge conflicts that need to be resolved anyway, so can you try rebasing and see if the failure recurs?

File: modules/hal_nxp/usb/usb_dc_mcux.h (hunk 4 above)
Comment (mmahadevan108) [reply modules/hal_nxp/usb/usb_dc_mcux.h]: Thank you, I have made the change.

**Comment by mmahadevan108** [comment]
> Why limited to those two MCUs? What about the LPC55S16 which is also part of Zephyr and that supports USB? Or is that limited due to hardware availability?

This driver should work on LPC55S16 as well. I do not have a board to test, I will be happy to review if you can submit a PR.

**Comment by mmahadevan108** [comment]
@jfischer-no This PR is adding USB device support on some NXP platforms. Could you kindly review and let me know if you have any feedback.

**Comment by MaureenHelm** [comment]
Rebase the hal_nxp PR to fix the buildkite failure. See zephyrproject-rtos/hal_nxp#109

**Review by hakehuang — COMMENTED** [review]
LPC and RT685 testing PASS. but for some samples which use endpoint up to 4(LPC55S69)/6(RT685), does not support.

**Review by MaureenHelm — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
binding looks fine

**Pull request merged**